pkg/kubelet/cm/topologymanager/socketmask/socketmask.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package socketmask
+
+import (
+	"fmt"
+)
+
+//SocketMask interface allows hint providers to create SocketMasks for TopologyHints
+type SocketMask interface {
+	Add(sockets ...int) error
+	Remove(sockets ...int) error
+	And(masks ...SocketMask)
+	Or(masks ...SocketMask)
+	Clear()
+	Fill()
+	IsEqual(mask SocketMask) bool
+	IsEmpty() bool
+	IsSet(socket int) bool
+	String() string
+	Count() int
+	GetSockets() []int
+}
+
+type socketMask uint64
+
+//NewSocketMask creates a new SocketMask
+func NewSocketMask(sockets ...int) (SocketMask, error) {
+	s := socketMask(0)
+	err := (&s).Add(sockets...)
+	if err != nil {
+		return nil, err
+	}
+	return &s, nil
+}
+
+//Add adds the sockets with topology affinity to the SocketMask
+func (s *socketMask) Add(sockets ...int) error {
+	mask := *s
+	for _, i := range sockets {
+		if i < 0 || i >= 64 {
+			return fmt.Errorf("socket number must be in range 0-63")
+		}
+		mask |= 1 << uint64(i)
+	}
+	*s = mask
+	return nil
+}
+
+//Remove removes specified sockets from SocketMask
+func (s *socketMask) Remove(sockets ...int) error {
+	mask := *s
+	for _, i := range sockets {
+		if i < 0 || i >= 64 {
+			return fmt.Errorf("socket number must be in range 0-63")
+		}
+		mask &^= 1 << uint64(i)
+	}
+	*s = mask
+	return nil
+}
+
+//And performs and operation on all bits in masks
+func (s *socketMask) And(masks ...SocketMask) {
+	for _, m := range masks {
+		*s &= *m.(*socketMask)
+	}
+}
+
+//Or performs or operation on all bits in masks
+func (s *socketMask) Or(masks ...SocketMask) {
+	for _, m := range masks {
+		*s |= *m.(*socketMask)
+	}
+}
+
+//Clear resets all bits in mask to zero
+func (s *socketMask) Clear() {
+	*s = 0
+}
+
+//Fill sets all bits in mask to one
+func (s *socketMask) Fill() {
+	*s = socketMask(^uint64(0))
+}
+
+//IsEmpty checks mask to see if all bits are zero
+func (s *socketMask) IsEmpty() bool {
+	return *s == 0
+}
+
+//IsSet checks socket in mask to see if bit is set to one
+func (s *socketMask) IsSet(socket int) bool {
+	if socket < 0 || socket >= 64 {
+		return false
+	}
+	return (*s & (1 << uint64(socket))) > 0
+}
+
+//IsEqual checks if masks are equal
+func (s *socketMask) IsEqual(mask SocketMask) bool {
+	return *s == *mask.(*socketMask)
+}
+
+//String converts mask to string
+func (s *socketMask) String() string {
```

Comment:
This String() method looks a bit strange, it will print binary representation of mask in opposite order compared to any canonical way of representing masks with lowest bit on the right.

Reply:
Agreed. This should probably be reversed.

Reply:
The sockets are set from left to right though - so it should be read that way as a string?

Reply:
The sockets are set as a bit in a uint64, which, if printed as a string, would have bit 0 on the far right and bit 63 on the far left. It's not a huge deal which order these are actually printed in the string representation -- I think @kad's point was just that if we are printing them to look like a binary 64-bit integer, why are we putting the most significant bit on the wrong side.

Reply:
@klueska 
```
return fmt.Sprintf("%064b", *s)
```
will be simple implementation of that function.

Reply:
This issue is being addressed in https://github.com/kubernetes/kubernetes/pull/78587